### PR TITLE
[SyncVM] Adding stdlib implementation to enable constant folding at E…

### PR DESCRIPTION
**Pull request by PavelKopyl**
[SyncVM] Adding stdlib implementation to enable constant folding at EarlyCSE.

### CODE REVIEW

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+; CHECK-LABEL: @test_signextend2
+; CHECK: ret i256 32767
+
+  %res = call i256 @__signextend(i256 1, i256 32767) builtin
+  ret i256 %res
+}
+
+define i256 @test_signextend3() {
+; CHECK-LABEL: @test_signextend3
+; CHECK: ret i256 poison
```
Comment (akiramenai) [inline llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll]: Why did we define it as poison?

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+define i256 @test_signextend2() {
+; CHECK-LABEL: @test_signextend2
+; CHECK: ret i256 32767
+
+  %res = call i256 @__signextend(i256 1, i256 32767) builtin
+  ret i256 %res
+}
+
```
Comment (akiramenai) [inline llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll:107]: Can we add a test when `singnextend` changes the value because the initial one occupy more bites than `op1 * 8 + 7`?

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll (hunk 1 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll]: This is a good question and I think we need to discuss behavior of stdlib functions in corner cases (when passed undef, poison values) in more details.
It seems reasonable for me to return 'poison'. Here is is how I'm reasoning about it:
 1. 'Undef' value is passed as the first argument (size in bytes minus one of the value being extended)
 2. We can replace 'Undef' with any determined value, including values that are greater that 31 (maximum size value)
 3. If the size is, for example, equal to 32 - signextend operation is not defined (though I'm not 100% sure about this, but I didn't find anything about this in EVM documentation), so the poison should be returned.
 
Please, correct me if I'm wrong. This is quite subtle topic.

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll (hunk 2 above)
Comment (PavelKopyl) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll:107]: Sure, will add.
Am I right you mean, for example, extension of 16-bit value '1000000011111111', passing '0' in the first argument, as if it was 8-bit size? In this case we should get (256)-1, as a result. Is that right?
I wonder, does the FE really generates such a code where we have inconsistency among the signextend arguments?

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll (hunk 2 above)
Comment (akiramenai) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll:107]: 1. That's right.
2. Anyone can write that in YUL insertion. As for YUL came from Solidity I doubt it ever generated.

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll (hunk 1 above)
Comment (akiramenai) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll]: Sounds reasonable. @vladimirradosavljevic, @lialan what do you think?

Comment (vladimirradosavljevic) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll]: This looks good to me, and taking into account trend of LLVM community to generate more poison values than undef, I think this is the way to go.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll (hunk 1 above)
Comment (lialan) [reply llvm/test/CodeGen/SyncVM/constant-folding-earlycse.ll]: The community is actually trying to get rid of undef and in favor of poison values. We should follow everybody else.

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 3]
```diff
@@ -2436,6 +2451,71 @@ static Constant *evaluateCompare(const APFloat &Op1, const APFloat &Op2,
...
 
+// SyncVM local begin
+/// Extending length of two’s complement signed integer.
+/// NumByte: the size in bytes minus one of the integer to sign extend.
+/// Val: the integer being extended.
+static Constant *ConstantFoldSignextendCall(Type *Ty, const APInt &NumByte,
+                                            const APInt &Val) {
+  unsigned BitWidth = Ty->getIntegerBitWidth();
+  APInt NumBits = NumByte * 8;
+  NumBits += 7;
```
Comment (akiramenai) [inline llvm/lib/Analysis/ConstantFolding.cpp]: Why not to use `NumByte * 8 + 7`. It's shorter and easier to read. `const` qualifier would also reinforce reader's confidence that it's a mere constant.

File: llvm/lib/Analysis/ConstantFolding.cpp [hunk 4]
```diff
@@ -2436,6 +2451,71 @@ static Constant *evaluateCompare(const APFloat &Op1, const APFloat &Op2,
...
+  if (Base == 0)
+    return Constant::getNullValue(Ty);
+  if (Base == 1)
+    return ConstantInt::get(Ty, APInt(BitWidth, 1));
+  if (Base == 2 && Exp.ule(255))
+    return ConstantInt::get(Ty, APInt(BitWidth, 1).shl(Exp));
+
+  // Target algorithm corresponds to the following C code,
+  // where m = 2 ** 256.
+  // long long binpow(long long a, long long b, long long m) {
```
Comment (akiramenai) [inline llvm/lib/Analysis/ConstantFolding.cpp:2489]: long long -> i256. Otherwise might cause wrong associations and confuse.

**Review by akiramenai — APPROVED** [review]
LGTM

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 4 above)
Comment (PavelKopyl) [reply llvm/lib/Analysis/ConstantFolding.cpp:2489]: Thanks, done.

File: llvm/lib/Analysis/ConstantFolding.cpp (hunk 3 above)
Comment (PavelKopyl) [reply llvm/lib/Analysis/ConstantFolding.cpp]: Fixed.

**Comment by PavelKopyl** [comment]
Please, note there is a difference in the calculated results by ConstantFolding and stdlib when Undef/Poision values are passed. But talking into account Undef/Poision nature that difference seems tom be fine (see constant-folding-earlycse.ll):
 - @test_addmod5
   CF: 0; STDLIB: poision
 - @test_signextend3
   CF: poison; STDLIB: -1
 - @test_exponent5
   CF: 0; STDLIB: Undef

**Comment by PavelKopyl** [comment]
One more thing. In order to eliminate  a call instruction itself, FE should generate declarations of the Stdlib functions with the attributes: { mustprogress nounwind readnone willreturn}

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```

**Comment by zksync-admin-bot2** [comment]
Benchmark results:
```
╔═╡ Size (-%) ╞═════════════════════╡ All ╞═╗
║ Mean                                0.006 ║
║ Best                               14.498 ║
║ Worst                               0.000 ║
║ Total                               0.011 ║
╠═╡ Cycles (-%) ╞═══════════════════╡ All ╞═╣
║ Mean                                0.006 ║
║ Best                               29.577 ║
║ Worst                              -3.806 ║
║ Total                               0.003 ║
╠═╡ Ergs (-%) ╞═════════════════════╡ All ╞═╣
║ Mean                                0.003 ║
║ Best                               17.190 ║
║ Worst                              -0.893 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

╔═╡ Size (-%) ╞═══════════════╡ Real life ╞═╗
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Cycles (-%) ╞═════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╠═╡ Ergs (-%) ╞═══════════════╡ Real life ╞═╣
║ Mean                                0.000 ║
║ Best                                0.000 ║
║ Worst                               0.000 ║
║ Total                               0.000 ║
╚═══════════════════════════════════════════╝

```
